### ISSUE TITLE
Service defaults not working for factories

### ISSUE THREAD

**Issue by curry684** [symfony/symfony#22143]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | dev-master

Given the new features with anonymous services and service defaults, I was assuming the following syntax would work:
```
services:
    _defaults:
        factory:        [ "@doctrine", getRepository ]

    MyProject\UserRepository:
        arguments:      [ "AppBundle:User" ]
```
This errors with: *"Please add the class to service "_defaults" even if it is constructed by a factory since we might need to add method calls based on compile-time checks."*

Changing the defaults to:
```
    _defaults:
        class:          Doctrine\ORM\EntityRepository
        factory:        [ "@doctrine", getRepository ]
```
Appears to work at first, but doesn't, as this actually causes a service called `_defaults` to be created, and the other services in the file to ignore the defaults altogether.

**Comment by curry684** [comment]
As a sidenote: shouldn't we completely be disallowing creation of a service called `_defaults` or `_instanceof` whatsoever? Or do we consider that BC breakage which it highly theoretically is?

**Comment by ro0NL** [comment]
>  Or do we consider that BC breakage which it highly theoretically is?

Yes :disappointed: 
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php#L328

The whole `_defaults`, `_instanceof` etc. services are ignored if we assume it's a regular service;
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php#L296

As of 4.0 this could be possible, and with `_instanceof` it will be possible. It should work with XML though...

**Comment by curry684** [comment]
Error part closed by #22185, feature part superseded by #22189
